Modify edge bolos

buckysballs opened the pull request:
Implementation of changes needed for #735 and transitively #671. Still need unit and full (wallet -> LB -> node) integration tests

marcinwadon commented on wallet/src/main/scala/org/constellation/wallet/Utils.scala:
```diff
@@ -158,7 +180,7 @@ object SignHelp {
 
   def signedObservationEdge(oe: ObservationEdge)(implicit kp: KeyPair): SignedObservationEdge =
-    SignedObservationEdge(hashSignBatchZeroTyped(Hashable.hash(oe), kp))
+    SignedObservationEdge(hashSignBatchZeroTyped(oe.hash, kp))
```
Why? I thought we should remove all implicit usages of kryo instead of adding new ones.

buckysballs replied:
Forgot 🤦‍♂

marcinwadon commented on wallet/src/main/scala/org/constellation/wallet/Utils.scala:
```diff
@@ -120,7 +134,15 @@ case class TypedEdgeHash(hash: String, hashType: EdgeHashType, baseHash: Option[
-)
+){
+  def hash = Hashing.sha256().hashBytes(KryoSerializer.serializeAnyRef(getRunLengthEncoding)).toString
+  def getRunLengthEncoding = {
```
`getRunLengthEncoding` should be abstracted somehow. Either in a separate trait or ideally I'd see an adapter for the hardware wallet which will be able to convert our structure into the hardware wallet compatible one.

Also, the "frame" should be tested and typed if possible. Now, you're able to return everything, in various order, which is very error prone.

buckysballs replied:
We should probably add it to Signable. We can replace hash there too

marcinwadon replied:
Yes, you can put a function declaration into a `Signable` and use it in a hash fn. It should be fine for now ;)

marcinwadon commented on src/main/scala/package.scala:
```diff
@@ -113,7 +113,7 @@ package object constellation extends POWExt with SignHelpExt with KeySerializeJS
 
   implicit class KryoSerExt(anyRef: AnyRef) {
-
+    def kryo(ref: AnyRef) = KryoSerializer.serializeAnyRef(ref)
```
Why it's needed? 🤔 We should limit the usage of such implicit dependencies as much as possible

buckysballs commented:
I removed the serialization implicits and moved everything into the Signable trait. There's one thing that we probably should do within the scope of this task, which is change our use of TypedEdgeHash to only pass is the run length encoding of whatever we wish to hash/sign, and let hashSignBatchZeroTyped take care of signing

marcinwadon commented:
I have no idea what happened with this PR but can we cherry-pick your commit (I think it's `0b1f543f21db3c4e947303e4ef2870bb7d5fb0fa`) and re-create a PR?

buckysballs commented:
It's super old, and yeah will do when we want to merge it. It's for the changes to the payload for encryption that we'll need for the hardware wallet. I just did it to knock it out

buckysballs commented:
Needs to fix for serialization across wallet and main project. Also need to calculate the hash of transaction and check whether it is equal to the signed hash in validateSourceSignature

marcinwadon commented on build.sbt:
```diff
@@ -180,7 +180,7 @@ lazy val root = (project in file("."))
     mainClass := Some("org.constellation.ConstellationNode")
     // other settings here
-  )
+  ).aggregate(wallet, keytool)
```
Why? It's not needed IMO

buckysballs replied:
Oh that just automatically makes jars for all sub-projects in the aggregate args, I'm lazy and didn't wanna compile them all individually. Should we remove for any reason?

buckysballs replied:
Re: dm about slow compile time, will remove

marcinwadon commented on src/main/scala/org/constellation/primitives/Schema.scala:
```diff
@@ -194,7 +195,15 @@ object Schema {
+      val parentValuesString = parents.flatMap(parent => parent.hashReference.length :: parent.hashReference :: Nil).mkString("")
+      val runEncodedTxData = data.hashReference//todo call this something else? Payload?
+        numParentsLengthString :: parentValuesString :: runEncodedTxData :: Nil mkString ""
+    }
```
Considering above function, it will look like this:
```
runLengthEncoding(parents.map(_.hashReference) ++ List(data.hashReference): _*)
```

buckysballs replied:
Yeah and I'm using Cons anyhow

marcinwadon commented on src/test/scala/org/constellation/TransactionHashIntegrityTest.scala:
```diff
@@ -1,26 +1,17 @@
-import org.http4s.client.dsl._
-import org.http4s.Method._
-import pl.abankowski.httpsigner.http4s.Http4sRequestVerifier
-import pl.abankowski.httpsigner.signature.generic.GenericVerifier
```
:+1: sorry, I've been testing sth and forgotten about remove

buckysballs replied:
Nbd

buckysballs commented on src/main/scala/org/constellation/domain/transaction/LastTransactionRef.scala:
```diff
@@ -1,9 +1,16 @@
 package org.constellation.domain.transaction
+import org.constellation.wallet.Signable
```
Ooo good catch

buckysballs replied:
Aaaaaand Ironically that's what was causing the serialization unit test to fail, great catch thanks man!

buckysballs replied:
Aaaaand I spoke too soon🤦‍♂️

buckysballs commented on src/main/scala/org/constellation/domain/transaction/LastTransactionRef.scala:
```diff
@@ -1,9 +1,16 @@
   ordinal: Long
-)
+) extends Signable  {
+  override def getRunLengthEncoding = {
```
I like this way better

buckysballs commented:
Last commit includes changes for #996 as well, for Transaction

buckysballs commented on src/main/scala/org/constellation/domain/transaction/TransactionValidator.scala:
```diff
@@ -11,7 +11,7 @@ object TransactionValidator {
   def validateSourceSignature(tx: Transaction): ValidationResult[Transaction] = {
     val isValid = tx.signatures.exists { hs ⇒
-      hs.publicKey.address == tx.src.address && hs.valid(tx.signaturesHash)
+      hs.publicKey.address == tx.src.address && hs.valid(tx.signaturesHash) && tx.hash == tx.signaturesHash //todo use a recalculation of get run length for OE and edge.signedObservationEdge.signatureBatch.hash
```
Note to remove todo

buckysballs commented:
Closing as all commits contained in pr #1057
